crates/oxc_linter/src/rules/eslint/no_undef.rs

```diff
@@ -0,0 +1,182 @@
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{Atom, Span};
+use oxc_syntax::operator::UnaryOperator;
+
+use crate::{context::LintContext, globals::BUILTINS, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-undef): Disallow the use of undeclared variables")]
+#[diagnostic(severity(warning), help("'{0}' is not defined."))]
+struct NoUndefDiagnostic(Atom, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUndef {
+    #[allow(dead_code)]
+    type_of: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow the use of undeclared variables.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It is most likely a potential ReferenceError caused by a misspelling of a variable or parameter name.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// var foo = someFunction();
+    /// var bar = a + 1;
+    /// ```
+    NoUndef,
+    correctness
+);
+
+impl Rule for NoUndef {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let type_of = value
+            .get(0)
+            .and_then(|config| config.get("typeof"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or_default();
+        Self { type_of }
+    }
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::IdentifierReference(ident) = node.kind() {
```

Comment:
@Boshen what should we do about type symbols? I don't think we bind any of typescript's `lib.d.ts` files to the top-level scope

Reply:
We don't lint `.d.ts` files.

Reply:
Correct, I'm talking about using them as an "environment", the same way tsc provides "built-in" types